content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+}
+----
+<1> This looks suspicious!
+
+We now would have to decide whether this image attribute should actually be loaded for this particular use case,
+(if so, we'd have to adjust the test accordingly),
+or whether it would for instance make more sense to mark this property as a lazily loaded one and only retrieve it when actually required.
+
+Solely working with the raw socket read and write events can be a bit cumbersome, though.
+Wouldn't it be nice if we also has the actual SQL statement which caused this I/O?
+Glad you asked!
+Neither Hibernate nor the Postgres JDBC driver emit any JFR events at the moment,
+but in part two of this blog post series, we'll discuss how to instrument an existing library to emit events like this,
+using a Java agent, without modifying the library in question.
+
+== Discussion and Summary
+
+JDK Flight Recorder and JfrUnit open up a very interesting approach for identifying potential performance regressions in Java applications.
+Instead of directly measuring an application's performance metrics,
+most notably latency and throughput,
+the idea is to measure and assert metrics that _impact_ the performance characteristics.
+This allows you to implement stable and reliable automated performance unit tests,
+whose outcome does not depend on the capabilities of the execution environment (e.g. number/size of CPUs),
+or other influential factors like concurrently running programs.
+
+Regressions in such impacting metrics, e.g. the amount of allocated memory, or bytes read from an application's database,
+are indicators that the application's performance may have been degraded.
+This approach offers some interesting advantages over performance tests on actual latency and throughput themselves:
+
+* *Hardware independent:* You can identify potential regressions also when running tests on hardware which is different (i.e. less powerful) from the actual production hardware
+* *Fast feedback cycle:* Being able to run performance regression tests on developer laptops, even in the IDE, allows for fast identification of potential regressions right during development, instead of having to wait for the results of less frequently executed test runs in a traditional performance test lab environment
+* *Robustness:* Tests are robust and not prone to factors such as the load induced by parallel jobs of a CI server or a virtualized/containerized environment
+* *Pro-active idenfication of performance issues:* Asserting a metric like memory allocation can help to identify future performance problems before they actual materialize; while the additional allocation rate may make no difference with the system's load as of today, it may negatively impact latency and throughput as the system reaches its limits with increased load; being able to identify the increased allocation rate early on allows for a more efficient handling of the situation while working on the code, compared to when finding out about such regression only later on
```

Comment:
```suggestion
* *Pro-active identification of performance issues:* Asserting a metric like memory allocation can help to identify future performance problems before they actual materialize; while the additional allocation rate may make no difference with the system's load as of today, it may negatively impact latency and throughput as the system reaches its limits with increased load; being able to identify the increased allocation rate early on allows for a more efficient handling of the situation while working on the code, compared to when finding out about such regression only later on
```

Reply:
Also might want to consider: in a complete production system there might be more components being integrated then the ones we have in our testbeds.

When measuring allocation rate of components individually, they might not seem problematic as it's easier to keep the allocation budget within the cheaper TLAB range, so while their individual allocation rate might not seem to have a problematic impact on throughput measurements performed on simple benchmarks, an excessive use of allocations could still translate in problematic bottlenecks on a more complex system as the same TLAB region is shared with other libraries.

Reply:
That's an interesting one; any idea how you'd identify such bottleneck? I don't think it needs discussion in this post (it's super long already), but would like to better understand and see what perhaps even can be done in JfrUnit itself towards that end.